### PR TITLE
Refactor functions in string solver

### PR DESCRIPTION
**Pull request by romainbrenguier**
This is to use standard algorithm and simplify the code

### CODE REVIEW

**Review by LAJW — APPROVED** [review]
Beautiful!

File: src/solvers/refinement/string_refinement.cpp [hunk 1]
```diff
@@ -2012,31 +2012,16 @@ static optionalt<exprt> find_counter_example(
...
-class gather_indices_visitort: public const_expr_visitort
+static array_index_mapt gather_indices(const exprt &expr)
 {
-public:
   array_index_mapt indices;
-
-  gather_indices_visitort(): indices() {}
-
-  void operator()(const exprt &expr) override
+  std::for_each(expr.depth_begin(), expr.depth_end(), [&](const exprt &expr)
```
Comment (allredj) [inline src/solvers/refinement/string_refinement.cpp]: The indentation is misleading. It should be clear that the lambda is entirely inside the `for_each` call.

File: src/solvers/refinement/string_refinement.cpp [hunk 2]
```diff
@@ -2025,31 +2025,17 @@ static array_index_mapt gather_indices(const exprt &expr)
...
-  is_linear_arithmetic_expr_visitort v;
-  expr.visit(v);
-  return v.correct;
+  auto non_linear_expr=std::find(
+    expr.depth_begin(),
+    expr.depth_end(),
+    [](const exprt &e)
+    {
+      return !e.operands().empty()
+        && (e.id()==ID_plus || e.id()==ID_minus || e.id()==ID_unary_minus);
```
Comment (allredj) [inline src/solvers/refinement/string_refinement.cpp]: Is that stating that a plus (or minus) expression is non_linear?

File: src/solvers/refinement/string_refinement.cpp [hunk 3]
```diff
@@ -2047,33 +2047,14 @@ static bool is_linear_arithmetic_expr(const exprt &expr)
...
-    // Inspect current value
-    const valuet cur=stack.top();
-    stack.pop();
-    const exprt e=cur.first;
-    const unsigned index_depth=cur.second;
-    const unsigned child_index_depth=index_depth+(e.id()==ID_index?0:1);
-
-    // If we found the universal variable not in an index_exprt, fail
-    if(e==expr.univ_var() && index_depth==0)
+  for(auto it=expr.depth_begin(); it!=expr.depth_end();)
```
Comment (allredj) [inline src/solvers/refinement/string_refinement.cpp]: That doesn't seem to do the same thing as the previous implementation. Does it still return true if the univ var is deeply nested in the body?

File: src/solvers/refinement/string_refinement.cpp [hunk 4]
```diff
@@ -2047,7 +2047,7 @@ static bool is_linear_arithmetic_expr(const exprt &expr)
 ///   false otherwise.
 static bool universal_only_in_index(const string_constraintt &expr)
 {
-  for(auto it=expr.depth_begin(); it!=expr.depth_end();)
+  for(auto it=expr.body().depth_begin(); it!=expr.body().depth_end();)
```
Comment (allredj) [inline src/solvers/refinement/string_refinement.cpp]: Could you please squash this with the previous commit? I still have doubts that it is correct. Can we look at it together?

**Review by allredj — CHANGES_REQUESTED** [review]

File: src/solvers/refinement/string_refinement.cpp (hunk 1 above)
Comment (LAJW) [reply src/solvers/refinement/string_refinement.cpp]: I disagree. Indentation is exactly where it should be. The only misleading bit is opening curly brace on a new line - the usual place for it is on the end of the previous line. Unfortunately CBMC linter doesn't allow that kind of indentation.

File: src/solvers/refinement/string_refinement.cpp [hunk 5]
```diff
@@ -2003,31 +2003,15 @@ static optionalt<exprt> find_counter_example(
...
-};
-
-/// \related string_constraintt
 static array_index_mapt gather_indices(const exprt &expr)
 {
-  gather_indices_visitort v;
-  expr.visit(v);
-  return v.indices;
+  array_index_mapt indices;
+  std::for_each(expr.depth_begin(), expr.depth_end(), [&](const exprt &expr) { // NOLINT
```
Comment (allredj) [inline src/solvers/refinement/string_refinement.cpp]: I still find that indentation hard to read. But if I'm the only one in that case then it's fine.

**Review by allredj — APPROVED** [review]

File: src/solvers/refinement/string_refinement.cpp (hunk 5 above)
Comment (romainbrenguier) [reply src/solvers/refinement/string_refinement.cpp]: @allredj I disabled clang-format on these lines to have brackets on a new line. Can you check it looks ok?

**Pull request merged**